src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs

```diff
@@ -113,41 +120,34 @@ private void SetEntry(CacheEntry entry)
             // Initialize the last access timestamp at the time the entry is added
             entry.LastAccessed = utcNow;
 
-            var added = false;
             CacheEntry priorEntry;
-
-            _entryLock.EnterWriteLock();
-            try
+            if (_entries.TryRemove(entry.Key, out priorEntry))
```

Comment:
Instead of TryRemove and TryAdd, we can use TryGet and TryUpdate. That should avoid a concurrent read reading a value of null.